### PR TITLE
[SPARK-42330][SQL] Assign the name `RULE_ID_NOT_FOUND` to the error class `_LEGACY_ERROR_TEMP_2175`

### PR DESCRIPTION
**Pull request by kori73**
### What changes were proposed in this pull request?

- Rename _LEGACY_ERROR_TEMP_2175 as RULE_ID_NOT_FOUND

- Add a test case for the error class.

### Why are the changes needed?

We are migrating onto error classes

### Does this PR introduce _any_ user-facing change?

Yes, the error message will include the error class name

### How was this patch tested?

`testOnly *RuleIdCollectionSuite` and Github Actions

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/rules/RuleIdCollectionSuite.scala [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+package org.apache.spark.sql.catalyst.rules
+
+import org.apache.spark.{SparkException, SparkFunSuite}
+
+class RuleIdCollectionSuite extends SparkFunSuite {
+
+    test("SPARK-42330: rule id not found") {
+        checkError(
+            exception = intercept[SparkException] {
+                RuleIdCollection.getRuleId("incorrect")
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/rules/RuleIdCollectionSuite.scala:27]: Can you trigger it from user code?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/rules/RuleIdCollectionSuite.scala [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.rules
+
+import org.apache.spark.{SparkException, SparkFunSuite}
+
+class RuleIdCollectionSuite extends SparkFunSuite {
+
+    test("SPARK-42330: rule id not found") {
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/rules/RuleIdCollectionSuite.scala:24]: Could you put it to `QueryExecutionErrorsSuite`, please.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/rules/RuleIdCollectionSuite.scala [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        checkError(
+            exception = intercept[SparkException] {
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/rules/RuleIdCollectionSuite.scala:26]: Could you fix indentation here and around. Please, follow the guide: https://github.com/databricks/scala-style-guide#indent

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/rules/RuleIdCollectionSuite.scala (hunk 1 above)
Comment (kori73) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/rules/RuleIdCollectionSuite.scala:27]: As far as I can understand this error is thrown when someone adds a `Rule` to apache spark source code and uses a method that requires `ruleId` such as follows:

[Analyzer.scala#L379](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L379-L380)

```
  object ResolveBinaryArithmetic extends Rule[LogicalPlan] {
    override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUpWithPruning(
      _.containsPattern(BINARY_ARITHMETIC), ruleId) { ... }
```

Then the developer will need to add this rule to `RuleIdCollection.rulesNeedingIds`.

[RuleIdCollection.scala](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleIdCollection.scala).

However, I can't see any way that a user can add an element to `RuleIdCollection.rulesNeedingIds`.

I might try to extend the Catalyst by injecting a rule which uses a method requiring `ruleId`. But the user will not be able to fix the error as described in the error message:

`Rule id not found for <ruleName>. Please modify RuleIdCollection.scala if you are adding a new rule.`

So, I think this error class/message is intended for spark developers although I might be completely wrong.

Also pinging the author @sigmod

Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/rules/RuleIdCollectionSuite.scala:27]: ok, let's assign an error class.

**Comment by MaxGekk** [comment]
@kori73 Could you resolve conflicts and address comments, please.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/rules/RuleIdCollectionSuite.scala (hunk 2 above)
Comment (kori73) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/rules/RuleIdCollectionSuite.scala]: moved to `QueryExecutionErrorsSuite`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/rules/RuleIdCollectionSuite.scala (hunk 3 above)
Comment (kori73) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/rules/RuleIdCollectionSuite.scala]: fixed indentation

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/rules/RuleIdCollectionSuite.scala (hunk 1 above)
Comment (kori73) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/rules/RuleIdCollectionSuite.scala]: added to error class doc

File: common/utils/src/main/resources/error/error-classes.json [hunk 4]
```diff
@@ -2423,6 +2423,12 @@
     ],
     "sqlState" : "42883"
   },
+  "RULE_ID_NOT_FOUND" : {
+    "message" : [
+      "Rule id not found for <ruleName>. Please modify RuleIdCollection.scala if you are adding a new rule."
```
Comment (MaxGekk) [inline common/utils/src/main/resources/error/error-classes.json:2428]: Let's highlight the rule name in the error message as we do for other parameters, and use the default quoting:
```suggestion
      "Not found an id for the rule name \"<ruleName>\". Please modify RuleIdCollection.scala if you are adding a new rule."
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 5]
```diff
@@ -1674,7 +1674,7 @@ private[sql] object QueryExecutionErrors extends QueryErrorsBase {
 
   def ruleIdNotFoundForRuleError(ruleName: String): Throwable = {
     new SparkException(
-      errorClass = "_LEGACY_ERROR_TEMP_2175",
+      errorClass = "RULE_ID_NOT_FOUND",
       messageParameters = Map(
         "ruleName" -> ruleName),
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:1679]:
```suggestion
      messageParameters = Map("ruleName" -> ruleName),
```

**Comment by MaxGekk** [comment]
@kori73 Please, update PR's title to:
```
[SPARK-42330][SQL] Assign the name `RULE_ID_NOT_FOUND` to the error class `_LEGACY_ERROR_TEMP_2175`
```

**Review by MaxGekk — APPROVED** [review]
Waiting for CI.

**Comment by kori73** [comment]
Thank you for your effort in reviewing

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master/3.5.
Thank you, @kori73.

**Pull request closed without merging**